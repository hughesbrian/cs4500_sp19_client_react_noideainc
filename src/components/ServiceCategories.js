import React from 'react'

const ServiceCategories = ({page, count, totalPages, serviceCategories,
                            serviceCategory, updateForm, createServiceCategory,
                            deleteServiceCategory, selectServiceCategory,
                            updateServiceCategory, setPageSelect,
                            setPagePagination}) =>
    <div>
        <h3>Service Categories</h3>
        <table className="table">
            <tbody>
            <tr>
                <td>
                    <b>Title</b>
                </td>
            </tr>
            <tr>
                <td>
                    <input
                        className="title-field"
                        onChange={e => updateForm(e)}
                        value={serviceCategory.title}></input>
                </td>
                <td>
                    <button className="btn btn-primary create-btn" onClick={createServiceCategory}>Add</button>
                    <button className="btn btn-success update-btn" onClick={updateServiceCategory}>Update</button>
                </td>
            </tr>
            {
                serviceCategories.map(sc =>
                    <tr key={sc.id} className="service-category-rows">
                        <td>{sc.title}</td>
                        <td><button className="btn btn-danger delete-btn" onClick={() => deleteServiceCategory(sc.id)}>
                            Delete
                        </button>
                        <button className="btn btn-primary select-btn" onClick={() => selectServiceCategory(sc.id)}>
                            Edit
                        </button>
                        </td>
                    </tr>
                )
            }
            <tr>
                <td>
                    <div className="row">
                        <div className="col-3">
                            <select
                                id="select-page"
                                value={page}
                                onChange={setPageSelect}
                                className="form-control">
                                {
                                    [...Array(totalPages).keys()].map(x =>
                                        <option className="page-select-fields" key={x} value={x}>{x + 1}</option>
                                    )
                                }
                            </select>
                        </div>
                        <div className="col-9">
                            <nav aria-label="Page navigation example">
                                <ul className="pagination">
                                    <li className="page-item">
                                        <button className="page-link previous-btn"
                                                onClick={() => setPagePagination(page - 1)}>
                                            Previous
                                        </button>
                                    </li>
                                    {
                                        [...Array(totalPages).keys()].map(x =>
                                            <li key={x} className="page-item">
                                                <button className="page-link pagination-btns"
                                                        onClick={() => setPagePagination(x)}>
                                                    {x + 1}
                                                 </button>
                                            </li>
                                        )
                                    }
                                    <li className="page-item">
                                        <button className="page-link next-btn"
                                                onClick={() => setPagePagination(page + 1)}>
                                            Next
                                        </button>
                                    </li>
                                </ul>
                            </nav>
                        </div>
                    </div>
                </td>
                <td>
                    {/* just for alignment */}
                </td>
            </tr>
            </tbody>
        </table>
    </div>

export default ServiceCategories
