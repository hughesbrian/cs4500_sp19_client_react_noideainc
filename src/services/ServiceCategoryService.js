const baseURL = "https://cs4500-sp19-noideainc.herokuapp.com";
//const baseURL = "http://localhost:8080";

export default class ServiceCategoryService {
    static instance = null;
    static getInstance() {
        if (ServiceCategoryService.instance === null) {
            ServiceCategoryService.instance = new ServiceCategoryService()
        }
        return this.instance
    }

    findServiceCategoryById = categoryId =>
        fetch(baseURL + `/api/categories/${categoryId}`)
            .then(response => response.json())

    findAllServiceCategories = () =>
        fetch(baseURL +  `/api/categories`)
            .then(response => response.json())

    pagedServiceCategories = (page, count) =>
        fetch(baseURL + `/api/categories/paged?page=${page}&count=${count}`)
            .then(response => response.json())

    createServiceCategory = sc => {
        delete sc.id;
        return fetch(baseURL + `/api/categories`, {
            method: 'post',
            body: JSON.stringify(sc),
            headers: {
                'content-type': 'application/json'
            }
        })
        .then(response => response.json())
    }

    updateServiceCategory = sc =>
        fetch(baseURL + `/api/categories/${sc.id}`, {
            method: 'put',
            body: JSON.stringify(sc),
            headers: {
                'content-type': 'application/json'
            }
        }).then(response => response.json())
    
    updateServiceCategoryScore = sc =>
        fetch(baseURL + `/api/categories/score/${sc.id}`, {
            method: 'put',
            body: JSON.stringify(sc),
            headers: {
                'content-type': 'application/json'
            }
        }).then(response => response.json())

    deleteServiceCategory = id =>
        fetch(baseURL + `/api/categories/${id}`, {
            method: 'delete'
        })

    findAllServicesByCategoryName = catName =>
        fetch(baseURL +  `/api/${catName}`)
            .then(response => response.json())
}
