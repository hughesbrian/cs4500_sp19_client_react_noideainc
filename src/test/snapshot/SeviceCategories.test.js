import React from 'react'
import ServiceCategories from '../../components/ServiceCategories'
import TestRenderer from 'react-test-renderer'
import serviceCategories from '../MockData/ServiceCategories.mock.json'
import ServiceCategoryService from '../../services/ServiceCategoryService'
const serviceCategoryService = ServiceCategoryService.getInstance()
import '../mockservice/ServiceCategoryService.mock'

test('create, delete, and update service categories', () => {
    const createServiceCategory = () => {
        serviceCategories.push({
            id: serviceCategories.length + 1,
            title: "title-" + serviceCategories.length + 1
        })

        let tree = testRenderer.toJSON()
        expect(tree).toMatchSnapshot()
    }

    const deleteServiceCategory = id => {
        serviceCategories.splice(0, 1)

        let tree = testRenderer.toJSON()
        expect(tree).toMatchSnapshot()
    }

    const updateServiceCategory = () => {
        serviceCategories[0] = {"id": serviceCategories[0].id, "title": "new title"}

        let tree = testRenderer.toJSON()
        expect(tree).toMatchSnapshot()
    }

    const testRenderer = TestRenderer.create(
        <ServiceCategories page={0}
                           count={serviceCategories.length}
                           totalPages={1}
                           serviceCategories={serviceCategories}
                           serviceCategory={serviceCategories[0]}
                           updateForm={() => null}
                           createServiceCategory={createServiceCategory}
                           deleteServiceCategory={deleteServiceCategory}
                           selectServiceCategory={() => null}
                           updateServiceCategory={updateServiceCategory}
                           setPageSelect={() => null}
                           setPagePagination={() => null}/>)

    let tree = testRenderer.toJSON()
    expect(tree).toMatchSnapshot()

    const testInstance = testRenderer.root

    const createBtn = testInstance.findByProps({className: 'btn btn-primary create-btn'})
    createBtn.props.onClick()

    const deleteBtns = testInstance.findAllByProps({className: 'btn btn-danger delete-btn'})
    deleteBtns[2].props.onClick()

    const updateBtn = testInstance.findByProps({className: 'btn btn-success update-btn'})
    updateBtn.props.onClick()
})

test('render list for service categories', () => {
    serviceCategoryService
        .findAllServiceCategories()
        .then(serviceCategories => {
            const testRenderer = TestRenderer.create(
                <ServiceCategories page={0}
                                   count={serviceCategories.length}
                                   totalPages={1}
                                   serviceCategories={serviceCategories}
                                   serviceCategory={serviceCategory[1]}
                                   updateForm={() => null}
                                   createServiceCategory={() => null}
                                   deleteServiceCategory={() => null}
                                   selectServiceCategory={() => null}
                                   updateServiceCategory={() => null}
                                   setPageSelect={() => null}
                                   setPagePagination={() => null}/>)

            let tree = testRenderer.toJSON()
            expect(tree).toMatchSnapshot()

            const serviceCategoriesRows = testInstance.findAllByProps({className: 'service-category-rows'})
            const selectBtns = testInstance.findAllByProps({className: 'btn btn-primary select-btn'})
            const createBtn = testInstance.findByProps({className: 'btn btn-primary create-btn'})
            const updateBtn = testInstance.findByProps({className: 'btn btn-success update-btn'})
            const deleteBtns = testInstance.findAllByProps({className: 'btn btn-danger delete-btn'})
            const previousBtn = testInstance.findByProps({className: 'page-link previous-btn'})
            const nextBtn = testInstance.findByProps({className: 'page-link next-btn'})
            const paginationBtns = testInstance.findAllByProps({className: 'page-link pagination-btns'})
            const titleField = testInstance.findByProps({className: 'title-field'})
            const pageSelectFields = testInstance.findByProps({className: 'page-select-fields'})

            expect(serviceCategoriesRows.props.length).toBe(3)
            expect(selectBtns.props.length).toBe(3)
            expect(createBtn.props.length).toBe(1)
            expect(updateBtn.props.length).toBe(1)
            expect(deleteBtns.props.length).toBe(3)
            expect(previousBtn.props.length).toBe(1)
            expect(nextBtn.props.length).toBe(1)
            expect(paginationBtns.props.length).toBe(1)
            expect(titleField.props.value).toBe("Fitness Trainer")
            expect(pageSelectFields.props.length).toBe(1)
        })
})
