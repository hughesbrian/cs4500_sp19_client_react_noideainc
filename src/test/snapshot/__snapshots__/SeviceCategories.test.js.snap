// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create, delete, and update service categories 1`] = `
<div>
  <h3>
    Service Categories
  </h3>
  <table
    className="table"
  >
    <tbody>
      <tr>
        <td>
          <b>
            Title
          </b>
        </td>
      </tr>
      <tr>
        <td>
          <input
            className="title-field"
            onChange={[Function]}
            value="Tutoring"
          />
        </td>
        <td>
          <button
            className="btn btn-primary create-btn"
            onClick={[Function]}
          >
            Add
          </button>
          <button
            className="btn btn-success update-btn"
            onClick={[Function]}
          >
            Update
          </button>
        </td>
      </tr>
      <tr
        className="service-category-rows"
      >
        <td>
          Tutoring
        </td>
        <td>
          <button
            className="btn btn-danger delete-btn"
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="btn btn-primary select-btn"
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
      </tr>
      <tr
        className="service-category-rows"
      >
        <td>
          Fitness Trainer
        </td>
        <td>
          <button
            className="btn btn-danger delete-btn"
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="btn btn-primary select-btn"
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
      </tr>
      <tr
        className="service-category-rows"
      >
        <td>
          House Painter
        </td>
        <td>
          <button
            className="btn btn-danger delete-btn"
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="btn btn-primary select-btn"
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
      </tr>
      <tr>
        <td>
          <div
            className="row"
          >
            <div
              className="col-3"
            >
              <select
                className="form-control"
                id="select-page"
                onChange={[Function]}
                value={0}
              >
                <option
                  className="page-select-fields"
                  value={0}
                >
                  1
                </option>
              </select>
            </div>
            <div
              className="col-9"
            >
              <nav
                aria-label="Page navigation example"
              >
                <ul
                  className="pagination"
                >
                  <li
                    className="page-item"
                  >
                    <button
                      className="page-link previous-btn"
                      onClick={[Function]}
                    >
                      Previous
                    </button>
                  </li>
                  <li
                    className="page-item"
                  >
                    <button
                      className="page-link pagination-btns"
                      onClick={[Function]}
                    >
                      1
                    </button>
                  </li>
                  <li
                    className="page-item"
                  >
                    <button
                      className="page-link next-btn"
                      onClick={[Function]}
                    >
                      Next
                    </button>
                  </li>
                </ul>
              </nav>
            </div>
          </div>
        </td>
        <td />
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`create, delete, and update service categories 2`] = `
<div>
  <h3>
    Service Categories
  </h3>
  <table
    className="table"
  >
    <tbody>
      <tr>
        <td>
          <b>
            Title
          </b>
        </td>
      </tr>
      <tr>
        <td>
          <input
            className="title-field"
            onChange={[Function]}
            value="Tutoring"
          />
        </td>
        <td>
          <button
            className="btn btn-primary create-btn"
            onClick={[Function]}
          >
            Add
          </button>
          <button
            className="btn btn-success update-btn"
            onClick={[Function]}
          >
            Update
          </button>
        </td>
      </tr>
      <tr
        className="service-category-rows"
      >
        <td>
          Tutoring
        </td>
        <td>
          <button
            className="btn btn-danger delete-btn"
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="btn btn-primary select-btn"
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
      </tr>
      <tr
        className="service-category-rows"
      >
        <td>
          Fitness Trainer
        </td>
        <td>
          <button
            className="btn btn-danger delete-btn"
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="btn btn-primary select-btn"
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
      </tr>
      <tr
        className="service-category-rows"
      >
        <td>
          House Painter
        </td>
        <td>
          <button
            className="btn btn-danger delete-btn"
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="btn btn-primary select-btn"
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
      </tr>
      <tr>
        <td>
          <div
            className="row"
          >
            <div
              className="col-3"
            >
              <select
                className="form-control"
                id="select-page"
                onChange={[Function]}
                value={0}
              >
                <option
                  className="page-select-fields"
                  value={0}
                >
                  1
                </option>
              </select>
            </div>
            <div
              className="col-9"
            >
              <nav
                aria-label="Page navigation example"
              >
                <ul
                  className="pagination"
                >
                  <li
                    className="page-item"
                  >
                    <button
                      className="page-link previous-btn"
                      onClick={[Function]}
                    >
                      Previous
                    </button>
                  </li>
                  <li
                    className="page-item"
                  >
                    <button
                      className="page-link pagination-btns"
                      onClick={[Function]}
                    >
                      1
                    </button>
                  </li>
                  <li
                    className="page-item"
                  >
                    <button
                      className="page-link next-btn"
                      onClick={[Function]}
                    >
                      Next
                    </button>
                  </li>
                </ul>
              </nav>
            </div>
          </div>
        </td>
        <td />
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`create, delete, and update service categories 3`] = `
<div>
  <h3>
    Service Categories
  </h3>
  <table
    className="table"
  >
    <tbody>
      <tr>
        <td>
          <b>
            Title
          </b>
        </td>
      </tr>
      <tr>
        <td>
          <input
            className="title-field"
            onChange={[Function]}
            value="Tutoring"
          />
        </td>
        <td>
          <button
            className="btn btn-primary create-btn"
            onClick={[Function]}
          >
            Add
          </button>
          <button
            className="btn btn-success update-btn"
            onClick={[Function]}
          >
            Update
          </button>
        </td>
      </tr>
      <tr
        className="service-category-rows"
      >
        <td>
          Tutoring
        </td>
        <td>
          <button
            className="btn btn-danger delete-btn"
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="btn btn-primary select-btn"
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
      </tr>
      <tr
        className="service-category-rows"
      >
        <td>
          Fitness Trainer
        </td>
        <td>
          <button
            className="btn btn-danger delete-btn"
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="btn btn-primary select-btn"
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
      </tr>
      <tr
        className="service-category-rows"
      >
        <td>
          House Painter
        </td>
        <td>
          <button
            className="btn btn-danger delete-btn"
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="btn btn-primary select-btn"
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
      </tr>
      <tr>
        <td>
          <div
            className="row"
          >
            <div
              className="col-3"
            >
              <select
                className="form-control"
                id="select-page"
                onChange={[Function]}
                value={0}
              >
                <option
                  className="page-select-fields"
                  value={0}
                >
                  1
                </option>
              </select>
            </div>
            <div
              className="col-9"
            >
              <nav
                aria-label="Page navigation example"
              >
                <ul
                  className="pagination"
                >
                  <li
                    className="page-item"
                  >
                    <button
                      className="page-link previous-btn"
                      onClick={[Function]}
                    >
                      Previous
                    </button>
                  </li>
                  <li
                    className="page-item"
                  >
                    <button
                      className="page-link pagination-btns"
                      onClick={[Function]}
                    >
                      1
                    </button>
                  </li>
                  <li
                    className="page-item"
                  >
                    <button
                      className="page-link next-btn"
                      onClick={[Function]}
                    >
                      Next
                    </button>
                  </li>
                </ul>
              </nav>
            </div>
          </div>
        </td>
        <td />
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`create, delete, and update service categories 4`] = `
<div>
  <h3>
    Service Categories
  </h3>
  <table
    className="table"
  >
    <tbody>
      <tr>
        <td>
          <b>
            Title
          </b>
        </td>
      </tr>
      <tr>
        <td>
          <input
            className="title-field"
            onChange={[Function]}
            value="Tutoring"
          />
        </td>
        <td>
          <button
            className="btn btn-primary create-btn"
            onClick={[Function]}
          >
            Add
          </button>
          <button
            className="btn btn-success update-btn"
            onClick={[Function]}
          >
            Update
          </button>
        </td>
      </tr>
      <tr
        className="service-category-rows"
      >
        <td>
          Tutoring
        </td>
        <td>
          <button
            className="btn btn-danger delete-btn"
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="btn btn-primary select-btn"
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
      </tr>
      <tr
        className="service-category-rows"
      >
        <td>
          Fitness Trainer
        </td>
        <td>
          <button
            className="btn btn-danger delete-btn"
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="btn btn-primary select-btn"
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
      </tr>
      <tr
        className="service-category-rows"
      >
        <td>
          House Painter
        </td>
        <td>
          <button
            className="btn btn-danger delete-btn"
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="btn btn-primary select-btn"
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
      </tr>
      <tr>
        <td>
          <div
            className="row"
          >
            <div
              className="col-3"
            >
              <select
                className="form-control"
                id="select-page"
                onChange={[Function]}
                value={0}
              >
                <option
                  className="page-select-fields"
                  value={0}
                >
                  1
                </option>
              </select>
            </div>
            <div
              className="col-9"
            >
              <nav
                aria-label="Page navigation example"
              >
                <ul
                  className="pagination"
                >
                  <li
                    className="page-item"
                  >
                    <button
                      className="page-link previous-btn"
                      onClick={[Function]}
                    >
                      Previous
                    </button>
                  </li>
                  <li
                    className="page-item"
                  >
                    <button
                      className="page-link pagination-btns"
                      onClick={[Function]}
                    >
                      1
                    </button>
                  </li>
                  <li
                    className="page-item"
                  >
                    <button
                      className="page-link next-btn"
                      onClick={[Function]}
                    >
                      Next
                    </button>
                  </li>
                </ul>
              </nav>
            </div>
          </div>
        </td>
        <td />
      </tr>
    </tbody>
  </table>
</div>
`;
